fc_util/src/arg_parser.rs

```diff
@@ -0,0 +1,542 @@
+    user_value: Option<String>,
+}
+
+impl<'a> ArgInfo<'a> {
+    pub fn new(name: &'a str) -> ArgInfo<'a> {
+        ArgInfo {
+            name,
+            required: false,
+            requires: None,
+            takes_value: false,
+            default_value: None,
+            optional: false,
+            help: None,
+            user_value: None,
+        }
+    }
+
+    pub fn required(mut self, required: bool) -> ArgInfo<'a> {
+        self.required = required;
+        self
+    }
+
+    pub fn requires(mut self, requires: &'a str) -> ArgInfo<'a> {
+        self.requires = Some(requires);
+        self
+    }
+
+    pub fn takes_value(mut self, takes_value: bool) -> ArgInfo<'a> {
+        self.takes_value = takes_value;
+        self
+    }
+
+    pub fn default_value(mut self, default_value: &'a str) -> ArgInfo<'a> {
+        self.default_value = Some(default_value);
+        self
+    }
+
+    pub fn optional(mut self, optional: bool) -> ArgInfo<'a> {
+        self.optional = optional;
+        self
+    }
+
+    pub fn help(mut self, help: &'a str) -> ArgInfo<'a> {
+        self.help = Some(help);
+        self
+    }
+
+    pub fn set_user_value(mut self, value: String) -> ArgInfo<'a> {
+        self.user_value = Some(value);
+        self
+    }
+
+    pub fn format_help(&self) -> String {
+        if let Some(help) = self.help {
+            if self.takes_value {
+                format!("--{} <{}>: {}", self.name, self.name, help)
+            } else {
+                format!("--{}: {}", self.name, help)
+            }
+        } else if self.takes_value {
+            format!("--{} <{}>", self.name, self.name)
+        } else {
+            format!("--{}", self.name)
+        }
+    }
+}
+
+impl<'a> PartialEq for ArgInfo<'a> {
+    fn eq(&self, other: &Self) -> bool {
+        self.name == other.name
+    }
+}
+
+impl<'a> Hash for ArgInfo<'a> {
+    fn hash<H: Hasher>(&self, state: &mut H) {
+        self.name.hash(state);
+    }
+}
+
+impl<'a> Borrow<str> for ArgInfo<'a> {
+    fn borrow(&self) -> &str {
+        self.name
+    }
+}
+
+#[derive(Debug, Default, PartialEq)]
+pub struct ArgParser<'a> {
+    args_matches: HashSet<ArgInfo<'a>>,
+    extra_args: Vec<String>,
+}
+
+impl<'a> ArgParser<'a> {
+    pub fn arg(mut self, arg_info: ArgInfo<'a>) -> Self {
+        self.args_matches.insert(arg_info);
+        self
+    }
+
+    pub fn value(&self, arg_name: &'static str) -> Option<String> {
+        self.args_matches.get(arg_name).unwrap().user_value.clone()
+    }
+
+    pub fn extra_args(self) -> Vec<String> {
+        self.extra_args
+    }
+
+    pub fn parse(self) -> Result<Self> {
+        let mut args: Vec<String> = env::args().collect();
+        // Remove the first element of `args` as it should be the binary name.
+        args.remove(0);
+
+        self.get_matches(args)
+    }
+
+    // Populates the fields of ArgParser.
+    pub fn get_matches(mut self, mut args: Vec<String>) -> Result<Self> {
```

Comment:
I think it would be better to receive the args as a reference (`&mut args: Vec<String>`) or slice (`&[String]`)